### DISCUSSION TITLE
How does Actix know handler arguments?

### DISCUSSION THREAD

**Discussion by aunyks** [actix/actix-web#2542] [Q&A] [answered]
I've noticed that all route handler functions don't need to accept the same number or types of arguments.

How does Actix know what these arguments are and how to populate them when it receives a request? I assume this happens at compile time.

Thanks in advance!

**Comment by asonix** [comment]
I believe that's a result of this macro that implements Handler for a given Fn type: https://github.com/actix/actix-web/blob/master/src/handler.rs#L65

**Comment by robjtede** [comment, marked as answer]
It's just a trait implemented for functions of various arities. It is the same reason that `fn foo(arg: impl Trait)` arguments work except that we restrict all the fn arguments to implement `FromRequest`, too.

The only other interesting part is that we implement `FromRequest` for tuples up to the same arity which satisfies the `T: FromRequest` bound you'll see on `web::to`, for example.

So, we can take the fn arg types, describe them as a tuple, run `from_request` on the tuple, and get extracted fn args back to call your handler with.

**Reply by robjtede** [reply]
written this up in the documentation:
https://actix.rs/actix-web/actix_web/trait.Handler.html#how-do-handlers-receive-variable-numbers-of-arguments

**Comment by aunyks** [comment]
Thanks everyone!
